speechbrain/core.py

```diff
@@ -793,17 +753,24 @@ def __init__(  # noqa: C901
             )
 
         logger.info(
-            f"Gradscaler enabled: {gradscaler_enabled}. Using precision: {self.precision}."
+            f"Gradscaler enabled: {gradscaler_enabled}. Using training precision: {self.precision}."
         )
-        self.scaler = torch.cuda.amp.GradScaler(enabled=gradscaler_enabled)
-
-        self.use_amp = False
-        if self.device == "cpu" and self.precision == "bf16":
-            self.use_amp = True
-        elif "cuda" in self.device and self.precision in ["fp16", "bf16"]:
-            self.use_amp = True
-
-        if self.use_amp and self.checkpointer is not None:
+        logger.info(
+            f"Evaluation precision: {self.eval_precision}."
+        )
+        self.scaler = torch.GradScaler(self.device, enabled=gradscaler_enabled)
+        train_dtype = AMPConfig.from_name(self.precision).dtype
+        self.training_ctx = TorchAutocast(
+            enabled=self.precision == "fp16" or self.precision == "bf16", device_type=self.device, dtype=train_dtype
+        )
```

Comment:
Could we just make `TorchAutocast` a no-op if dtype == "fp32"?

Reply:
Done.